src/uarch.rs

```diff
@@ -0,0 +1,380 @@
+use crate::uarch::Core::{Heterogeneous, Homogenous};
+use crate::Vendor;
+use crate::Vendor::{Intel, Amd};
+
+#[non_exhaustive]
+pub enum CoreArch {
+    // Not including Intel micro-architecture without CPUID suport, for now.
+    // Intel Micro-architectures (with CPUID support)
+    i486,
+    P5,
+    P6,
+    NetBurst,
+    Prescott,
+    PentiumM,
+    ModifiedPentiumM,
+    Core,
+    Nehalem,
+    SandyBridge,
+    Haswell,
+    // Client Skylake Core are pretty much minor refreshes (follow wikipedia)
+    // However Server cores have actually undergone more evolution, so distinguish them
+    Skylake,
+    SkylakeServer,
+    CascadeLake,
+    CooperLake,
+    PalmCove,
+    SunnyCove,
+    WillowCove,
+    CypressCove,
+    GoldenCove,
+    // Intel Small Cores
+    Bonnel,
+    Saltwell,
+    Silvermont,
+    Airmont,
+    Goldmont,
+    GoldmontPlus,
+    Tremont,
+    Gracemont,
+    Crestmont,
+    // Other Intel
+    LakeMont,
+    // Many Core TODO
+
+    // AMD Micro-architectures (with CPUID support)
+    Am486,
+    EnhancedAm486,
+    Am586,
+    K5,
+    K6,
+    K6_2,
+    K6III,
+    K7,
+    K8, // Hammer
+    // K9 was never finished.
+    K10, // Llano
+    Bobcat,
+    Jaguar,
+    // Enhanced Jaguar for XboX One X 'Scorpio". (CPUID currently unknown)
+    Puma,
+    Bulldozer,
+    Piledriver,
+    SteamRoller,
+    Excavator,
+    // Zen,
+    Zen,
+    ZenPlus, // PinnacleRidge, Picasso,
+    Zen2,
+    Zen3,
+    Zen4,
+}
+
+#[non_exhaustive]
+pub enum Core {
+    Homogenous(CoreArch),
+    Heterogeneous { P: CoreArch, E: CoreArch },
```

Comment:
this could maybe just be `Vec<CoreArch>` or `&static [CoreArch]`? seems tricky to be sure there are only two cores especially in the future

Reply:
That was one of the options, I felt I was perhaps a bit more painful to handle, but I will make that change. I think ARM land has some CPUs with three types of core too.